src/GatewayAPI/Database/TransactionQuerier.cs

```diff
@@ -332,37 +344,113 @@ private async Task<List<long>> GetAccountTransactionStateVersions(AccountTransac
+        {
+            await using var localCtx = await _dbContextFactory.CreateDbContextAsync();
+
+            return await localCtx.Validators.Where(r => ids.Contains(r.Id)).ToDictionaryAsync(r => r.Id);
+        }
+
+        async Task<Dictionary<long, Account>> LoadAccounts(List<long> ids)
+        {
+            await using var localCtx = await _dbContextFactory.CreateDbContextAsync();
+
+            return await localCtx.Accounts.Where(r => ids.Contains(r.Id)).ToDictionaryAsync(r => r.Id);
+        }
+
+        async Task<Dictionary<byte[], RawTransaction>> LoadRawTransactions(List<byte[]> ids)
+        {
+            await using var localCtx = await _dbContextFactory.CreateDbContextAsync();
+
+            return await localCtx.RawTransactions.Where(r => ids.Contains(r.TransactionIdentifierHash)).ToDictionaryAsync(r => r.TransactionIdentifierHash, ByteArrayEqualityComparer.Default);
+        }
+
+        var transactionWithOperationGroups = await LoadLedgerTransactions(transactionStateVersions);
+
+        // Now you might be thinking... wtf?! And that would be a really good thinking of yours!
+        // We used to operate on single query (AsSplitQuery() technically) but database/entityframework cost
+        // associated with ~10k result sets was simply too big.
+        // What's more future version of this class will most likely cache ~99% of those resources in-memory.
+        var resourceIds = new HashSet<long>();
+        var validatorIds = new HashSet<long>();
+        var fromAccountIds = new HashSet<long>();
+        var toAccountIds = new HashSet<long>();
+        var rawTransactionIds = new HashSet<byte[]>(ByteArrayEqualityComparer.Default);
+
+        foreach (var x in transactionWithOperationGroups)
+        {
+            rawTransactionIds.Add(x.TransactionIdentifierHash);
+
+            foreach (var y in x.SubstantiveOperationGroups)
+            {
+                if (y.InferredAction == null)
+                {
+                    continue;
+                }
+
+                resourceIds.Add(y.InferredAction.ResourceId ?? -1);
+                validatorIds.Add(y.InferredAction.ValidatorId ?? -1);
+                fromAccountIds.Add(y.InferredAction.FromAccountId ?? -1);
+                toAccountIds.Add(y.InferredAction.ToAccountId ?? -1);
+            }
+        }
+
+        var resourcesTask = LoadResources(resourceIds.Where(id => id > 0).ToList());
+        var validatorsTask = LoadValidators(validatorIds.Where(id => id > 0).ToList());
+        var accountsTask = LoadAccounts(fromAccountIds.Concat(toAccountIds).Where(id => id > 0).ToList());
+        var rawTransactionsTask = LoadRawTransactions(rawTransactionIds.ToList());
+
+        await Task.WhenAll(resourcesTask, validatorsTask, accountsTask, rawTransactionsTask);
+
+        var rh = new ResourceHolder(await resourcesTask, await validatorsTask, await accountsTask, await rawTransactionsTask);
```

Comment:
Would it make sence to check the result status? Task can be cancelled or faulted....

Reply:
`await` will rethrow if any of those tasks has been cancelled/failed.